<?xml version="1.0"?>

<!--
License: Apache 2.0. See LICENSE file in root directory.
Copyright(c) 2017 Intel Corporation. All Rights Reserved

This is the URDF model for the Intel RealSense 410 camera, in it's
aluminum peripherial evalution case.
-->

<robot name="sensor_r410" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:property name="M_PI" value="3.1415926535897931" />

  <!-- The following values are approximate, and the camera node
   publishing TF values with actual calibrated camera extrinsic values -->
  <xacro:property name="r410_cam_depth_to_left_ir_offset" value="-0.0275"/>
  <xacro:property name="r410_cam_depth_to_right_ir_offset" value="0.0275"/>

  <!-- The following values model the aluminum peripherial case for the
    R410 camera, with the camera joint represented by the actual 
    peripherial camera tripod mount -->
  <xacro:property name="r410_cam_width" value="0.099"/>
  <xacro:property name="r410_cam_height" value="0.035"/>
  <xacro:property name="r410_cam_depth" value="0.0185"/>
  <xacro:property name="r410_cam_mount_from_center_offset" value=".0375"/>

  <!-- The following offset is relative the the physical R410 camera peripherial
    camera tripod mount -->
  <xacro:property name="r410_cam_depth_px" value="0.00"/>
  <xacro:property name="r410_cam_depth_py" value="-0.0425"/>
  <xacro:property name="r410_cam_depth_pz" value="0.026"/>

  <material name="aluminum">
    <color rgba="0.5 0.5 0.5 1"/>
  </material>

  <xacro:macro name="sensor_r410" params="parent *origin">

    <!-- camera body, with origin at bottom screw mount -->
    <joint name="camera_joint" type="fixed">
      <xacro:insert_block name="origin" />
      <parent link="${parent}"/>
      <child link="camera_link" />
    </joint>

    <link name="camera_link">
      <visual>
      <origin xyz="0 ${-r410_cam_mount_from_center_offset} ${r410_cam_height/2}" rpy="${M_PI/2} 0 ${M_PI/2}"/>
        <geometry>
          <box size="${r410_cam_width} ${r410_cam_height} ${r410_cam_depth}"/>
          <!--<mesh filename="package://realsense2_camera/meshes/r410/r410.dae" />-->
        </geometry>
        <material name="aluminum"/>
      </visual>
      <collision>
        <origin xyz="0.0 0.0 ${r410_cam_height/2}" rpy="0 0 0"/>
        <geometry>
        <box size="${r410_cam_depth} ${r410_cam_width} ${r410_cam_height}"/>
        </geometry>
      </collision>
      <inertial>
        <!-- The following are not reliable values, and should not be used for modeling -->
        <mass value="0.564" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.003881243" ixy="0.0" ixz="0.0" iyy="0.000498940" iyz="0.0" izz="0.003879257" />
      </inertial>
    </link>
   
    <!-- camera depth joints and links -->
    <joint name="camera_depth_joint" type="fixed">
      <origin xyz="${r410_cam_depth_px} ${r410_cam_depth_py} ${r410_cam_depth_pz}" rpy="0 0 0"/>
      <parent link="camera_link"/>
      <child link="camera_depth_frame" />
    </joint>
    <link name="camera_depth_frame"/>

    <joint name="camera_depth_optical_joint" type="fixed">
      <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
      <parent link="camera_depth_frame" />
      <child link="camera_depth_optical_frame" />
    </joint>
    <link name="camera_depth_optical_frame"/>
      
    <!-- camera left IR joints and links -->
    <joint name="camera_left_ir_joint" type="fixed">
      <origin xyz="0 ${r410_cam_depth_to_left_ir_offset} 0" rpy="0 0 0" />
      <parent link="camera_depth_frame" />
      <child link="camera_left_ir_frame" />
    </joint>
    <link name="camera_left_ir_frame"/>

    <joint name="camera_left_ir_optical_joint" type="fixed">
      <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
      <parent link="camera_left_ir_frame" />
      <child link="camera_left_ir_optical_frame" />
    </joint>
    <link name="camera_left_ir_optical_frame"/>

    <!-- camera right IR joints and links -->
    <joint name="camera_right_ir_joint" type="fixed">
      <origin xyz="0 ${r410_cam_depth_to_right_ir_offset} 0" rpy="0 0 0" />
      <parent link="camera_depth_frame" />
      <child link="camera_right_ir_frame" />
    </joint>
    <link name="camera_right_ir_frame"/>

    <joint name="camera_right_ir_optical_joint" type="fixed">
      <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
      <parent link="camera_right_ir_frame" />
      <child link="camera_right_ir_optical_frame" />
    </joint>
    <link name="camera_right_ir_optical_frame"/>
  </xacro:macro>
</robot>
